sound/soc/sof/ipc4-control.c

```diff
@@ -327,13 +329,26 @@ static int sof_ipc4_bytes_ext_put(struct snd_sof_control *scontrol,
 		return -EINVAL;
 	}
 
+	/* We need to restore data in any situation where the byte control update fails. Save it here. */
+	old_cdata = kmemdup(data, scontrol->max_size, GFP_KERNEL);
+	if (!old_cdata)
+		return -ENOMEM;
+
 	/* Copy the whole binary data which includes the ABI header and the payload */
-	if (copy_from_user(data, tlvd->tlv, header.length))
+	if (copy_from_user(data, tlvd->tlv, header.length)) {
+		memcpy(data, old_cdata, scontrol->max_size);
```

Comment:
is this step really necessary here?

Reply:
copy_from_user zeroes out the destination buffer if the copy fails because of -EFAULT, whether it succeeded in copying at least one byte or it didn't. So yes, a restore is needed.